packages/adapter/src/chains/moonbeam.ts

```diff
@@ -0,0 +1,231 @@
+import _ from 'lodash';
+import BN from 'bn.js';
+import type { SubmittableExtrinsic, AddressOrPair } from '@polkadot/api/types';
+import type { WeightV2 } from '@polkadot/types/interfaces';
+import type { u64, u128, Option } from '@polkadot/types';
+import type { HexString } from '@polkadot/util/types';
+import type { KeyringPair } from '@polkadot/keyring/types';
+import { Asset, ChainAsset, Weight } from '@oak-network/sdk-types';
+import { ChainAdapter, TaskScheduler } from './chainAdapter';
+import { convertAbsoluteLocationToRelative, getDerivativeAccountV3, sendExtrinsic } from '../util';
+import { AccountType, SendExtrinsicResult } from '../types';
+import { WEIGHT_REF_TIME_PER_SECOND } from '../constants';
+
+const TRANSACT_XCM_INSTRUCTION_COUNT = 4;
```

Comment:
If every chain contains this variable, it should be defined in the base class `chainAdapter` such as `this.defaultXcmInstructionCount` with a default value and overwritten by specific chainAdapters.

Besides, there’s no need to defined a get function for a variable. It should be retrievable by `oakAdapter.defaultXcmInstructionCount`

Reply:
In fact, `TRANSACT_XCM_INSTRUCTION_COUNT` is not a common member variable for every chain.

Only the **Adapter that implements the `TaskScheduler` interface needs this variable.

```
export interface TaskScheduler {
  /**
   * Get the instruction number of XCM instructions for transact
   */
  getTransactXcmInstructionCount(): number;
  
  /**
   * Schedule Task through XCM message
   * @param destination The location of the destination chain
   * @param encodedTaskExtrinsic encoded task extrinsic
   * @param feeLocation Fee location
   * @param feeAmount Fee amount
   * @param encodedCallWeight The encoded call weight weight of the XCM instructions
   * @param overallWeight The overall weight of the XCM instructions
   * @param keyringPair Operator's keyring pair
   */
  scheduleTaskThroughXcm(destination: any, encodedTaskExtrinsic: HexString, feeLocation: any, feeAmount: BN, encodedCallWeight: Weight, overallWeight: Weight, keyringPair: KeyringPair): Promise<SendExtrinsicResult>;
}
```

TaskScheduler defines the `scheduleTaskThroughXcm` function to enable a chain to send an XCM to Turing for scheduling a task.

`AstarAdapter` and `MoonbeamAdapter` implement `TaskScheduler`. However, `MangataAdapter` and `TuringAdapter` do not implement `TaskScheduler`.

To calculate the `overallWeight` parameter required for the `scheduleTaskThroughXcm` function, you need to know the number of instructions included in this XCM operation.

overallWeight = InstructionCountWillBeSentFromAstar * InstructionWeightOfTuring / WEIGHT_PER_SECOND

`TaskScheduler` defines the `getTransactXcmInstructionCount` function to return the number of instructions included in this XCM operation.

Therefore, this `TRANSACT_XCM_INSTRUCTION_COUNT` is actually only relevant to the implementation of the `scheduleTaskThroughXcm` operation in `TaskScheduler`.

For example, in the implementation of `xcmTransactor.transactThroughSigned` in Moonbeam, when the `refund` parameter is false, there are 5 XCM instructions, and when the refund parameter is true, there are 4 XCM instructions.

For example, in Astar, when we use `polkadotXcm.send`, we can construct 6 XCM instructions that include `RefundSubplus` and `DepositAsset`, or 4 XCM instructions that do not include them.

Therefore, I believe it should not be added as a common member variable to `ChainAdapter`.

Reply:
No, being a member variable in the base class **does not mean every inheritance has to use it**. It is a fact that every chainAdapter will need this instruction count for xcm, sooner or later, so it’s better to add it now in to chainAdapter.